core/cachemap.go

```diff
@@ -38,7 +39,19 @@ func NewCacheMap[K comparable, T any]() *CacheMap[K, T] {
 	}
 }
 
-func (m *CacheMap[K, T]) GetOrInitialize(key K, fn func() (T, error)) (T, error) {
+type cacheMapContextKey[K comparable, T any] struct {
+	key K
+	m   *CacheMap[K, T]
+}
+
+var ErrCacheMapRecursiveCall = fmt.Errorf("recursive call detected")
```

Comment:
@vito just want to make sure this isn't lost in the noise

I added this to enable detection of recursive module dependencies w/out deadlock. I originally added the detection using normal DAG walking approaches but then still hit deadlock due to the fact that module loading is wrapped in a `CacheMap` in `schema/schema.go`.

This should only ever detect situations that are 100% guaranteed to result in deadlock, so I think it's probably a good change to have in general too.

Reply:
Ha, I ran into a deadlock in the same place in #6158 so :+1: